# Project Presentation

My project will be looking at four search algorithms to find the solution to the 8-Puzzle *[eyt-puhz-uhl]*

### The Algorithms that will be looked at are:
* Breadth-First-Search (**BFS**)
* Depth-First-Iterative- Deepening Search (**DFID**)
* A-Star (**A***)
* Iterative-Deepening A Star (**IDA***)

### The topic I plan on exploring are:
* Comparing the nodes that are expanded/generated of all Algorithms
* Is there a difference in expanded/generated nodes produced by **BFS** and **DFID** ?
* Does **A*** ever outperform **IDA*** ?
* Is there a difference in expanded/generated nodes produced by **A*** and **IDA*** ?

## General Problem
The solver must be able to find a solution in the least amount of moves

       Start State                 Goal State
<img src="example.png" alt="drawing" width="400"/>

In [1]:
from helpers import *
from solvers import *
from generate_data import *
import figures

## What I have Completed
I have finished:
* Implementation of A*
* Have a Implementation IDA* ( *There Is a Issue* )

The Issue I have having with **IDA***is that there are some states that produce unoptimal solutions 

---

### The Issue of with My Implementation IDA*

In [2]:
# Get the Data
data = get_data([Astar,idastar],50,50,123456780).astype({'expanded':'int64','generated':'int64','sol_length':'int64','state':'int64'})
# Clean the Data
data['optimal'] = data['state'].apply(lambda x: len(data[data['state'] == x]['sol_length'].unique()) == 1)
data['ratio'] = data['generated'] / data['expanded']
data['time_ms'] = data.time * 1000

In [3]:
data[~data['optimal']]

index     name  expanded  generated  sol_length      time      state  \
6       0   A Star       699       1121          20  0.053417  236504817   
7       0  IDAStar       929       1512          22  0.049821  236504817   
80      0   A Star       388        627          18  0.029339  823416075   
81      0  IDAStar       907       1573          20  0.050545  823416075   

    optimal     ratio    time_ms  
6     False  1.603720  53.416729  
7     False  1.627557  49.820900  
80    False  1.615979  29.339075  
81    False  1.734289  50.544739

In [4]:
goal_state = int_to_matrix(123456780)
test_state = int_to_matrix(236504817)

        Test State
<img src="problem.png" alt="drawing" width="200"/>

In [5]:
solver1 = Astar(goal_state)
solver2 = idastar(goal_state)
sol1 = solver1.find_solution(test_state,get_location(test_state,0))
print("Generated ASTAR",solver1.num_of_generated)
print("Expanded ASTAR",solver1.num_of_expanded)
print("len of sol ASTAR", len(sol1) - 1)
sol2 = solver2.find_solution(test_state,get_location(test_state,0))
print("Generated IDA",solver2.num_of_generated)
print("Expanded IDA",solver2.num_of_expanded)
print("len of sol ida", len(sol2) - 1)
print(solver2.thresholds)

Generated ASTAR 1121
Expanded ASTAR 699
len of sol ASTAR 20
Generated IDA 1512
Expanded IDA 929
len of sol ida 22
{12: [9, 9], 14: [66, 50], 16: [244, 155], 18: [635, 401], 20: [1466, 906], 22: [1512, 929]}



---
For the Test State, **A*** *generates 1121* nodes and *expands 699* of them while **IDA*** *generates 1512* nodes and *expands 929*

**IDA*** should have found the solution at ***threshold = 20*** but my implementation does an additional iteration at ***threshold = 22***



### Some Figures comparing A* and IDA* (Botched version)

I will only look at solutions that are optimal

In [6]:
opt_data = data[data['optimal']]

In [9]:
# figures.ratio_plot(opt_data)
# figures.node_plot(opt_data)

<img src="ratio.jpg" alt="drawing" width="400"/>

*Note* I'm comparing the ratios of the 2 search algorithms

<img src="node_plot.png" alt="drawing"/>

## Finished